In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from pathlib import Path

import warnings
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots

from sklearn.preprocessing import StandardScaler, OneHotEncoder, MinMaxScaler
from sklearn.decomposition import PCA
from sklearn.cluster import KMeans, DBSCAN
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import silhouette_score, calinski_harabasz_score, davies_bouldin_score

from sklearn.preprocessing import StandardScaler
from sklearn.metrics import silhouette_score, calinski_harabasz_score, davies_bouldin_score
from sklearn.decomposition import PCA
import pandas as pd
import numpy as np

import hdbscan
from gower import gower_matrix

from utils import *

In [2]:
df_sintomas_disc = pd.read_excel(r"C:\Users\Lan_1\Documents\VictorL_TesinaMACI\VictorL_TesinaMACI\Python\data\Sintomas_disc_Infante_Juvenil_n=1558.xlsx")

In [3]:
full_dict = build_full_dict(df_sintomas_disc)

# Clusterización de sujetos por variables categóricas

Como ya se ha explicado, es necesario caracterizar las clases latentes de individuos. En este caso, se hará únicamente en base a las preguntas basales que hayan respondido. ¿Pero son estas preguntas realmente categóricas? Se hace una revisión recordando que para cada módulo, estas son las preguntas basales:

| Código | Padecimiento                                      | Basales                              | N |
|--------|--------------------------------------------------|--------------------------------------|---|
| pso    | Fobia social                                     | 1–5                                  | 5 |
| pga    | Ansiedad generalizada                            | 1–5, 23–29                           | 12 |
| psa    | Ansiedad por separación                          | 1–12                                 | 12 |
| pea    | Bulimia                                          | 1–4, 10–12, 20                       | 8 |
| pmd    | Depresión mayor o distimia                       | 1–22, 35                             | 23 |
| pma    | Manía o hipomanía                                | 1–13                                 | 13 |
| psz    | Esquizofrenia                                    | 1–2, 9–14                            | 8 |
| pad    | Trastorno por déficit de atención e hiperactividad | 1–2, 4–11, 22–25, 27–29, 31–33, 44 | 21 |
| pcd    | Trastorno de conducta                            | 1–4, 7–9, 12–15, 19–23, 25, 27–29, 39 | 21 |
| pod    | Trastorno de oposición desafiante                | 1–12                                 | 12 |
| pal    | Abuso de alcohol                                 | 1                                    | 1 |
| pmj    | Consumo de marihuana                             | 1                                    | 1 |
| psu    | Consumo de otras sustancias                      | 1–11                                 | 11 |
| pni    | Consumo de tabaco                                | 1–2                                  | 2 |


In [5]:
tol_report(full_dict, 3)

Agorafobia - pag - N=64


------------------------------------------------------------------------------------------------------------------------------------------------------


Fobia específica - psp - N=70


------------------------------------------------------------------------------------------------------------------------------------------------------


Fobia social - pso - N=55
Tolerancia de 10% => Umbral de 0.06 a 43.97


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,pct_88.0,pct_99.0,limited
0,pso001,0.064185,4.043646,95.892169,0.0,0.0,0.0,0.000000,0.0,0.0,True
2,pso002,0.064185,24.326059,75.609756,0.0,0.0,0.0,0.000000,0.0,0.0,True
3,pso003,0.064185,67.907574,32.028241,0.0,0.0,0.0,0.000000,0.0,0.0,True
5,pso004,0.064185,78.818999,20.924262,0.0,0.0,0.0,0.192555,0.0,0.0,True
6,pso005,0.064185,51.155327,48.652118,0.0,0.0,0.0,0.128370,0.0,0.0,True


Tolerancia de 13% => Umbral de 0.06 a 54.17


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,pct_88.0,pct_99.0,limited
7,pso006,43.966624,12.451861,43.388960,0.0,0.0,0.0,0.192555,0.0,0.0,True
8,pso007,43.966624,37.355584,18.549422,0.0,0.0,0.0,0.128370,0.0,0.0,True


Tolerancia de 24% => Umbral de 54.17 a 67.01


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,pct_88.0,pct_99.0,limited
9,pso008,54.172015,27.535302,18.292683,0.0,0.0,0.0,0.000000,0.0,0.0,True
10,pso009,54.172015,25.802311,20.025674,0.0,0.0,0.0,0.000000,0.0,0.0,True
11,pso010,54.172015,16.174583,29.589217,0.0,0.0,0.0,0.064185,0.0,0.0,True
12,pso011,54.172015,27.599487,18.164313,0.0,0.0,0.0,0.064185,0.0,0.0,True
13,pso012,54.172015,29.075738,16.688062,0.0,0.0,0.0,0.064185,0.0,0.0,True
14,pso013,54.172015,24.646983,21.181001,0.0,0.0,0.0,0.000000,0.0,0.0,True




------------------------------------------------------------------------------------------------------------------------------------------------------


Ansiedad generalizada - pga - N=103
Tolerancia de 11% => Umbral de 0.13 a 2.89


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,limited
5,pga002,0.12837,82.284981,17.394095,0.0,0.0,0.0,0.192555,True
10,pga003,0.12837,61.489089,38.189987,0.0,0.0,0.0,0.192555,True
15,pga004,0.12837,49.101412,50.577664,0.0,0.0,0.0,0.192555,True
20,pga005,0.12837,76.315789,23.363286,0.0,0.0,0.0,0.192555,True
72,pga023,0.12837,71.887035,27.920411,0.0,0.0,0.0,0.064185,True
76,pga024,0.12837,60.397946,39.345315,0.0,0.0,0.0,0.128370,True
79,pga025,0.12837,71.758665,27.984596,0.0,0.0,0.0,0.128370,True
82,pga026,0.12837,71.437741,28.241335,0.0,0.0,0.0,0.192555,True
85,pga027,0.12837,77.920411,21.822850,0.0,0.0,0.0,0.128370,True
90,pga028,0.12837,81.899872,17.843389,0.0,0.0,0.0,0.128370,True


Tolerancia de 12% => Umbral de 0.13 a 42.68


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,limited
0,pga001,2.888318,53.530167,43.324775,0.0,0.0,0.064185,0.192555,True


Tolerancia de 13% => Umbral de 0.13 a 49.42


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,limited
25,pga006,42.682927,32.156611,25.160462,0.0,0.0,0.0,0.0,True




------------------------------------------------------------------------------------------------------------------------------------------------------


Mutismo selectivo - psm - N=42


------------------------------------------------------------------------------------------------------------------------------------------------------


Pánico - ppa - N=128


------------------------------------------------------------------------------------------------------------------------------------------------------


Trastorno por estrés postraumático - ppt - N=90


------------------------------------------------------------------------------------------------------------------------------------------------------


Ansiedad por separación - psa - N=76
Tolerancia de 15% => Umbral de 0.06 a 2.82


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,pct_88.0,pct_99.0,limited
0,psa001,0.064185,74.903723,24.967908,0.0,0.0,0.0,0.064185,0.0,0.0,True
6,psa003,0.064185,90.949936,8.921694,0.0,0.0,0.0,0.064185,0.0,0.0,True
9,psa004,0.064185,76.636714,23.170732,0.0,0.0,0.0,0.128370,0.0,0.0,True
13,psa005,0.064185,62.708601,37.227214,0.0,0.0,0.0,0.000000,0.0,0.0,True
16,psa006,0.064185,37.227214,62.708601,0.0,0.0,0.0,0.000000,0.0,0.0,True
24,psa007,0.064185,48.780488,50.962773,0.0,0.0,0.0,0.192555,0.0,0.0,True
28,psa008,0.064185,75.288832,24.326059,0.0,0.0,0.0,0.320924,0.0,0.0,True
31,psa009,0.064185,82.605905,17.073171,0.0,0.0,0.0,0.256739,0.0,0.0,True
34,psa010,0.064185,79.460847,20.474968,0.0,0.0,0.0,0.000000,0.0,0.0,True
37,psa011,0.064185,13.414634,86.521181,0.0,0.0,0.0,0.000000,0.0,0.0,True


Tolerancia de 16% => Umbral de 0.06 a 13.48


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,pct_88.0,pct_99.0,limited
3,psa002,2.824134,58.472401,38.575096,0.0,0.0,0.0,0.12837,0.0,0.0,True


Tolerancia de 18% => Umbral de 0.06 a 37.29


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,pct_88.0,pct_99.0,limited
38,psa011a,13.478819,76.957638,9.563543,0.0,0.0,0.0,0.0,0.0,0.0,True




------------------------------------------------------------------------------------------------------------------------------------------------------


Trastorno obsesivo-compulsivo - poc - N=89


------------------------------------------------------------------------------------------------------------------------------------------------------


Bulimia - pea - N=117
Tolerancia de 6% => Umbral de 0.13 a 0.19


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,pct_77.0,pct_88.0,pct_99.0,limited
5,pea002k,0.12837,0.000000,0.000000,0.0,0.0,0.000000,0.000000,0.000000,0.000000,0.770218,False
7,pea003k,0.12837,0.000000,0.000000,0.0,0.0,0.000000,0.000000,0.064185,0.641849,1.219512,False
8,pea004,0.12837,70.731707,26.765083,0.0,0.0,0.000000,0.000000,0.000000,0.000000,0.000000,False
21,pea010,0.12837,75.930680,23.748395,0.0,0.0,0.000000,0.192555,0.000000,0.000000,0.000000,True
25,pea011,0.12837,76.251605,23.427471,0.0,0.0,0.000000,0.192555,0.000000,0.000000,0.000000,True
28,pea012,0.12837,87.419769,12.387677,0.0,0.0,0.000000,0.064185,0.000000,0.000000,0.000000,True
72,pea020,0.12837,24.582798,22.079589,0.0,0.0,53.145058,0.064185,0.000000,0.000000,0.000000,True


Tolerancia de 7% => Umbral de 0.13 a 50.13


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,pct_77.0,pct_88.0,pct_99.0,limited
3,pea001z,0.192555,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.898588,False


Tolerancia de 12% => Umbral de 50.13 a 51.67


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,pct_77.0,pct_88.0,pct_99.0,limited
35,pea013,50.12837,43.196406,6.611040,0.0,0.0,0.0,0.064185,0.0,0.0,0.0,True
39,pea014,50.12837,47.560976,2.246470,0.0,0.0,0.0,0.064185,0.0,0.0,0.0,True
45,pea015,50.12837,49.293967,0.513479,0.0,0.0,0.0,0.064185,0.0,0.0,0.0,True
51,pea016,50.12837,48.973042,0.834403,0.0,0.0,0.0,0.064185,0.0,0.0,0.0,True
59,pea017,50.12837,47.111682,2.695764,0.0,0.0,0.0,0.064185,0.0,0.0,0.0,True
65,pea018,50.12837,46.726573,3.080873,0.0,0.0,0.0,0.064185,0.0,0.0,0.0,True




------------------------------------------------------------------------------------------------------------------------------------------------------


Trastorno de eliminación - pel - N=138


------------------------------------------------------------------------------------------------------------------------------------------------------


Pica - ppi - N=43


------------------------------------------------------------------------------------------------------------------------------------------------------


Trastorno de tics - ptc - N=57


------------------------------------------------------------------------------------------------------------------------------------------------------


Tricotilomanía - ptr - N=43


------------------------------------------------------------------------------------------------------------------------------------------------------


Depresión mayor o distimia - pmd - N=222
Tolerancia de 10% => Umbral de 0.13 a 5.58


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,pct_88.0,limited
0,pmd001,0.12837,57.060334,42.811297,0.0,0.0,0.0,0.000000,0.0,True
7,pmd002,0.12837,77.278562,22.593068,0.0,0.0,0.0,0.000000,0.0,True
11,pmd003,0.12837,51.347882,48.523748,0.0,0.0,0.0,0.000000,0.0,True
17,pmd004,0.12837,77.278562,22.528883,0.0,0.0,0.0,0.064185,0.0,True
23,pmd005,0.12837,74.454429,25.353017,0.0,0.0,0.0,0.064185,0.0,True
27,pmd006,0.12837,80.616175,19.191271,0.0,0.0,0.0,0.064185,0.0,True
31,pmd007,0.12837,71.630295,28.241335,0.0,0.0,0.0,0.000000,0.0,True
35,pmd008,0.12837,65.789474,34.082157,0.0,0.0,0.0,0.000000,0.0,True
40,pmd009,0.12837,73.234917,26.636714,0.0,0.0,0.0,0.000000,0.0,True
44,pmd010,0.12837,86.071887,13.735558,0.0,0.0,0.0,0.064185,0.0,True


Tolerancia de 11% => Umbral de 0.13 a 40.44


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,pct_88.0,limited
162,pmd035,5.584082,77.856226,16.495507,0.0,0.0,0.0,0.064185,0.0,True
169,pmd036,6.996149,73.812580,19.191271,0.0,0.0,0.0,0.000000,0.0,True


Tolerancia de 12% => Umbral de 0.13 a 51.48


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,pct_88.0,limited
45,pmd010a,40.436457,52.824134,6.611040,0.0,0.064185,0.0,0.064185,0.0,True
50,pmd011a,47.432606,45.057766,7.509628,0.0,0.000000,0.0,0.000000,0.0,True




------------------------------------------------------------------------------------------------------------------------------------------------------


Manía o hipomanía - pma - N=113
Tolerancia de 12% => Umbral de 99.55 a 99.87


,question,pct_nan,pct_0.0,pct_2.0,pct_8.0,limited
0,pma001,99.550706,0.320924,0.128370,0.000000,True
11,pma002,99.550706,0.385109,0.064185,0.000000,True
19,pma003,99.550706,0.449294,0.000000,0.000000,True
25,pma004,99.550706,0.449294,0.000000,0.000000,True
31,pma005,99.550706,0.449294,0.000000,0.000000,True
36,pma006,99.550706,0.320924,0.128370,0.000000,True
41,pma007,99.550706,0.449294,0.000000,0.000000,True
46,pma008,99.550706,0.385109,0.064185,0.000000,True
53,pma009,99.550706,0.385109,0.064185,0.000000,True
59,pma010,99.550706,0.385109,0.064185,0.000000,True


Tolerancia de 16% => Umbral de 99.55 a 99.94


,question,pct_nan,pct_0.0,pct_2.0,pct_8.0,limited
1,pma001a,99.87163,0.128370,0.000000,0.0,True
37,pma006a,99.87163,0.064185,0.064185,0.0,True
38,pma006b,99.87163,0.064185,0.064185,0.0,True
65,pma011a,99.87163,0.128370,0.000000,0.0,True
66,pma011b,99.87163,0.128370,0.000000,0.0,True


Tolerancia de 26% => Umbral de 99.94 a 100.00


,question,pct_nan,pct_0.0,pct_2.0,pct_8.0,limited
12,pma002a,99.935815,0.064185,0.000000,0.0,True
40,pma006d,99.935815,0.064185,0.000000,0.0,True
47,pma008a,99.935815,0.064185,0.000000,0.0,True
48,pma008b,99.935815,0.000000,0.064185,0.0,True
49,pma008c,99.935815,0.000000,0.064185,0.0,True
50,pma008d,99.935815,0.000000,0.064185,0.0,True
52,pma008f,99.935815,0.064185,0.000000,0.0,True
54,pma009a,99.935815,0.064185,0.000000,0.0,True
55,pma009b,99.935815,0.064185,0.000000,0.0,True
60,pma010a,99.935815,0.064185,0.000000,0.0,True




------------------------------------------------------------------------------------------------------------------------------------------------------


Esquizofrenia - psz - N=311
Tolerancia de 5% => Umbral de 0.13 a 56.61


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,pct_88.0,limited
0,psz001,0.12837,89.602054,10.077022,0.0,0.0,0.0,0.192555,0.0,True
8,psz002,0.12837,91.270860,8.472401,0.0,0.0,0.0,0.128370,0.0,True
49,psz009,0.12837,88.125802,11.617458,0.0,0.0,0.0,0.128370,0.0,True
59,psz010,0.12837,79.139923,20.603338,0.0,0.0,0.0,0.128370,0.0,True
69,psz011,0.12837,95.635430,4.107831,0.0,0.0,0.0,0.128370,0.0,True
79,psz012,0.12837,95.827985,3.915276,0.0,0.0,0.0,0.128370,0.0,True
88,psz013,0.12837,98.780488,0.962773,0.0,0.0,0.0,0.128370,0.0,True
99,psz014,0.12837,98.138639,1.604621,0.0,0.0,0.0,0.128370,0.0,True


Tolerancia de 7% => Umbral de 0.13 a 79.40


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,pct_88.0,limited
109,psz015,56.61104,42.811297,0.577664,0.0,0.0,0.0,0.0,0.0,True
119,psz016,56.61104,41.206675,2.182285,0.0,0.0,0.0,0.0,0.0,True
130,psz017,56.61104,41.591784,1.797176,0.0,0.0,0.0,0.0,0.0,True
139,psz018,56.61104,41.206675,2.182285,0.0,0.0,0.0,0.0,0.0,True
150,psz019,56.61104,42.554557,0.834403,0.0,0.0,0.0,0.0,0.0,True
161,psz020,56.61104,35.879332,7.509628,0.0,0.0,0.0,0.0,0.0,True
171,psz021,56.61104,43.003851,0.385109,0.0,0.0,0.0,0.0,0.0,True
181,psz022,56.61104,41.206675,2.182285,0.0,0.0,0.0,0.0,0.0,True
217,psz028,56.61104,40.500642,2.888318,0.0,0.0,0.0,0.0,0.0,True
226,psz029,56.61104,38.703466,4.685494,0.0,0.0,0.0,0.0,0.0,True


Tolerancia de 8% => Umbral de 56.61 a 91.53


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,pct_88.0,limited
1,psz001a,89.922978,4.749679,5.327343,0.0,0.0,0.0,0.0,0.0,True
50,psz009a,88.382542,6.546855,5.070603,0.0,0.0,0.0,0.0,0.0,True
60,psz010a,79.396662,15.340180,5.263158,0.0,0.0,0.0,0.0,0.0,True




------------------------------------------------------------------------------------------------------------------------------------------------------


Trastorno por déficit de atención e hiperactividad - pad - N=180
Tolerancia de 12% => Umbral de 0.13 a 10.27


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,pct_77.0,pct_88.0,pct_99.0,limited
0,pad001,0.12837,60.141207,39.666239,0.0,0.0,0.0,0.064185,0.0,0.0,0.0,True
5,pad002,0.12837,70.539153,29.204108,0.0,0.0,0.0,0.128370,0.0,0.0,0.0,True
15,pad004,0.12837,56.225931,43.581515,0.0,0.0,0.0,0.064185,0.0,0.0,0.0,True
20,pad005,0.12837,48.587933,51.091142,0.0,0.0,0.0,0.192555,0.0,0.0,0.0,True
25,pad006,0.12837,62.644416,37.098845,0.0,0.0,0.0,0.128370,0.0,0.0,0.0,True
30,pad007,0.12837,69.640565,30.102696,0.0,0.0,0.0,0.128370,0.0,0.0,0.0,True
35,pad008,0.12837,69.768935,29.910141,0.0,0.0,0.0,0.192555,0.0,0.0,0.0,True
40,pad009,0.12837,74.646983,25.096277,0.0,0.0,0.0,0.128370,0.0,0.0,0.0,True
45,pad010,0.12837,69.897304,29.845956,0.0,0.0,0.0,0.128370,0.0,0.0,0.0,True
53,pad011,0.12837,60.077022,39.666239,0.0,0.0,0.0,0.128370,0.0,0.0,0.0,True


Tolerancia de 13% => Umbral de 0.13 a 19.13


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,pct_77.0,pct_88.0,pct_99.0,limited
10,pad003,10.269576,63.157895,26.444159,0.0,0.0,0.0,0.12837,0.0,0.0,0.0,True
104,pad026,12.901155,65.275995,21.694480,0.0,0.0,0.0,0.12837,0.0,0.0,0.0,True


Tolerancia de 14% => Umbral de 0.13 a 56.42


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,pct_77.0,pct_88.0,pct_99.0,limited
21,pad005a,48.908858,18.100128,32.991014,0.0,0.0,0.0,0.00000,0.0,0.0,0.0,True
124,pad030,19.127086,67.522465,13.222080,0.0,0.0,0.0,0.12837,0.0,0.0,0.0,True




------------------------------------------------------------------------------------------------------------------------------------------------------


Trastorno de conducta - pcd - N=256
Tolerancia de 9% => Umbral de 0.13 a 1.86


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,pct_77.0,pct_88.0,pct_99.0,limited
0,pcd001,0.12837,85.109114,14.698331,0.0,0.0,0.000000,0.064185,0.0,0.0,0.0,True
1,pcd002,0.12837,95.635430,4.172015,0.0,0.0,0.000000,0.064185,0.0,0.0,0.0,True
2,pcd003,0.12837,92.747112,7.060334,0.0,0.0,0.000000,0.064185,0.0,0.0,0.0,True
3,pcd004,0.12837,99.229782,0.577664,0.0,0.0,0.000000,0.064185,0.0,0.0,0.0,True
13,pcd007,0.12837,99.550706,0.256739,0.0,0.0,0.000000,0.064185,0.0,0.0,0.0,True
14,pcd008,0.12837,99.550706,0.256739,0.0,0.0,0.000000,0.064185,0.0,0.0,0.0,True
15,pcd009,0.12837,99.550706,0.256739,0.0,0.0,0.000000,0.064185,0.0,0.0,0.0,True
23,pcd012,0.12837,90.115533,9.691913,0.0,0.0,0.000000,0.064185,0.0,0.0,0.0,True
31,pcd013,0.12837,96.790757,3.016688,0.0,0.0,0.000000,0.064185,0.0,0.0,0.0,True
40,pcd014,0.12837,85.494223,14.249037,0.0,0.0,0.000000,0.128370,0.0,0.0,0.0,True


Tolerancia de 10% => Umbral de 0.13 a 1.99


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,pct_77.0,pct_88.0,pct_99.0,limited
49,pcd018,1.861361,74.646983,23.427471,0.0,0.0,0.0,0.064185,0.0,0.0,0.0,True
120,pcd026,1.925546,87.548139,10.462131,0.0,0.0,0.0,0.064185,0.0,0.0,0.0,True
174,pcd029t,1.861361,96.662388,1.476252,0.0,0.0,0.0,0.000000,0.0,0.0,0.0,True


Tolerancia de 11% => Umbral de 0.13 a 76.57


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,pct_77.0,pct_88.0,pct_99.0,limited
194,pcd036,1.98973,94.672657,3.273427,0.0,0.0,0.0,0.064185,0.0,0.0,0.0,True
203,pcd037,1.98973,85.365854,12.580231,0.0,0.0,0.0,0.064185,0.0,0.0,0.0,True
204,pcd038,1.98973,87.034660,10.911425,0.0,0.0,0.0,0.064185,0.0,0.0,0.0,True




------------------------------------------------------------------------------------------------------------------------------------------------------


Trastorno de oposición desafiante - pod - N=101
Tolerancia de 12% => Umbral de 0.13 a 44.29


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,pct_88.0,pct_99.0,limited
0,pod001,0.12837,62.195122,37.548139,0.0,0.0,0.0,0.128370,0.0,0.0,True
5,pod002,0.12837,43.966624,55.712452,0.0,0.0,0.0,0.192555,0.0,0.0,True
10,pod003,0.12837,70.410783,29.332478,0.0,0.0,0.0,0.128370,0.0,0.0,True
15,pod004,0.12837,60.012837,39.730424,0.0,0.0,0.0,0.128370,0.0,0.0,True
20,pod005,0.12837,75.930680,23.812580,0.0,0.0,0.0,0.128370,0.0,0.0,True
27,pod006,0.12837,74.454429,25.288832,0.0,0.0,0.0,0.128370,0.0,0.0,True
34,pod007,0.12837,57.766367,42.105263,0.0,0.0,0.0,0.000000,0.0,0.0,True
39,pod008,0.12837,57.702182,42.105263,0.0,0.0,0.0,0.064185,0.0,0.0,True
46,pod009,0.12837,45.571245,54.107831,0.0,0.0,0.0,0.192555,0.0,0.0,True
53,pod010,0.12837,91.142490,8.600770,0.0,0.0,0.0,0.128370,0.0,0.0,True


Tolerancia de 13% => Umbral de 0.13 a 52.63


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,pct_88.0,pct_99.0,limited
6,pod002a,44.287548,25.545571,30.166881,0.0,0.0,0.0,0.0,0.0,0.0,True


Tolerancia de 14% => Umbral de 0.13 a 57.89


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,pct_88.0,pct_99.0,limited
49,pod009c,52.631579,23.748395,23.620026,0.0,0.0,0.0,0.0,0.0,0.0,True




------------------------------------------------------------------------------------------------------------------------------------------------------


Abuso de alcohol - pal - N=110
Tolerancia de 5% => Umbral de 0.13 a 88.19


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,pct_88.0,pct_99.0,limited
0,pal001,0.128370,75.288832,24.518614,0.0,0.0,0.000000,0.064185,0.000000,0.0,True
1,pal001ay,75.481386,0.000000,0.064185,0.0,0.0,0.320924,0.256739,0.064185,0.0,False
3,pal001b,79.011553,4.236200,16.752246,0.0,0.0,0.000000,0.000000,0.000000,0.0,True
4,pal001c,79.717587,8.472401,11.810013,0.0,0.0,0.000000,0.000000,0.000000,0.0,True
5,pal001d,87.291399,9.948652,2.759949,0.0,0.0,0.000000,0.000000,0.000000,0.0,True


Tolerancia de 13% => Umbral de 88.19 a 88.32


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,pct_88.0,pct_99.0,limited
7,pal001f,88.189987,0.000000,2.952503,1.668806,0.000000,0.000000,0.000000,0.0,0.000000,False
8,pal001g,88.189987,2.631579,9.178434,0.000000,0.000000,0.000000,0.000000,0.0,0.000000,True
10,pal001i,88.189987,0.000000,2.118100,0.962773,0.320924,0.320924,0.128370,0.0,0.000000,False
11,pal001j,88.189987,8.793325,0.641849,0.256739,0.192555,0.064185,0.000000,0.0,0.064185,False
12,pal001k,88.189987,7.381258,4.428755,0.000000,0.000000,0.000000,0.000000,0.0,0.000000,True
13,pal001ly,88.189987,2.374840,0.000000,0.000000,0.000000,0.000000,0.064185,0.0,0.064185,False
23,pal006,88.189987,10.462131,1.347882,0.000000,0.000000,0.000000,0.000000,0.0,0.000000,True
30,pal009,88.189987,10.654685,1.155327,0.000000,0.000000,0.000000,0.000000,0.0,0.000000,True
32,pal010,88.189987,10.975610,0.834403,0.000000,0.000000,0.000000,0.000000,0.0,0.000000,True


Tolerancia de 14% => Umbral de 88.19 a 88.96


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,pct_88.0,pct_99.0,limited
15,pal002,88.318357,10.526316,1.091142,0.0,0.0,0.064185,0.0,0.0,0.0,True




------------------------------------------------------------------------------------------------------------------------------------------------------


Consumo de marihuana - pmj - N=87
Tolerancia de 5% => Umbral de 0.13 a 96.02


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_8.0,pct_9.0,limited
0,pmj001,0.128370,93.453145,6.354300,0.0,0.0,0.064185,True
1,pmj001ay,93.645700,0.000000,0.000000,0.0,0.0,0.064185,False
3,pmj001b,94.544288,2.118100,3.337612,0.0,0.0,0.000000,True
4,pmj001c,95.763800,1.861361,2.374840,0.0,0.0,0.000000,True


Tolerancia de 6% => Umbral de 0.13 a 97.63


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_8.0,pct_9.0,limited
5,pmj001d,96.020539,3.016688,0.962773,0.0,0.0,0.0,True


Tolerancia de 22% => Umbral de 97.63 a 97.69


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_8.0,pct_9.0,limited
7,pmj001f,97.62516,0.000000,0.770218,0.256739,0.0,0.000000,False
8,pmj001g,97.62516,1.091142,1.283697,0.000000,0.0,0.000000,True
18,pmj006,97.62516,2.118100,0.256739,0.000000,0.0,0.000000,True
24,pmj009,97.62516,2.182285,0.192555,0.000000,0.0,0.000000,True
26,pmj010,97.62516,1.668806,0.706033,0.000000,0.0,0.000000,True
28,pmj011,97.62516,1.668806,0.706033,0.000000,0.0,0.000000,True
31,pmj012,97.62516,1.925546,0.449294,0.000000,0.0,0.000000,True
33,pmj013,97.62516,1.925546,0.449294,0.000000,0.0,0.000000,True
35,pmj014,97.62516,2.053915,0.320924,0.000000,0.0,0.000000,True
37,pmj015,97.62516,1.925546,0.385109,0.000000,0.0,0.064185,True




------------------------------------------------------------------------------------------------------------------------------------------------------


Consumo de otras sustancias - psu - N=240
Tolerancia de 5% => Umbral de 0.13 a 99.23


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,limited
0,psu001,0.12837,99.614891,0.192555,0.0,0.000000,0.0,0.064185,True
11,psu002,0.12837,99.550706,0.192555,0.0,0.064185,0.0,0.064185,True
22,psu003,0.12837,99.037227,0.770218,0.0,0.000000,0.0,0.064185,True
33,psu004,0.12837,99.807445,0.000000,0.0,0.000000,0.0,0.064185,True
43,psu005,0.12837,99.807445,0.000000,0.0,0.000000,0.0,0.064185,True
54,psu006,0.12837,99.807445,0.000000,0.0,0.000000,0.0,0.064185,True
64,psu007,0.12837,99.743261,0.064185,0.0,0.000000,0.0,0.064185,True
75,psu008,0.12837,99.807445,0.000000,0.0,0.000000,0.0,0.064185,True
86,psu009,0.12837,99.679076,0.128370,0.0,0.000000,0.0,0.064185,True
97,psu010,0.12837,99.743261,0.064185,0.0,0.000000,0.0,0.064185,True


Tolerancia de 6% => Umbral de 0.13 a 99.49


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,limited
23,psu003a,99.229782,0.770218,0.0,0.0,0.0,0.0,0.0,True
24,psu003by,99.229782,0.000000,0.0,0.0,0.0,0.0,0.0,False


Tolerancia de 7% => Umbral de 0.13 a 99.55


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,limited
26,psu003c,99.486521,0.320924,0.192555,0.0,0.0,0.0,0.0,True
28,psu003e,99.486521,0.513479,0.000000,0.0,0.0,0.0,0.0,True




------------------------------------------------------------------------------------------------------------------------------------------------------


Consumo de tabaco - pni - N=85
Tolerancia de 5% => Umbral de 0.13 a 83.89


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,pct_99.0,limited
0,pni001,0.128370,79.589217,20.089859,0.000000,0.000000,0.000000,0.192555,0.0,True
1,pni001ay,79.910141,0.000000,0.000000,0.064185,0.256739,0.256739,0.513479,0.0,False
3,pni001b,82.092426,5.584082,12.323492,0.000000,0.000000,0.000000,0.000000,0.0,True
6,pni002,0.128370,95.186136,0.064185,0.000000,0.192555,0.000000,4.428755,0.0,True


Tolerancia de 6% => Umbral de 0.13 a 85.49


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,pct_99.0,limited
67,pni028,83.889602,11.232349,4.878049,0.0,0.0,0.0,0.0,0.0,True


Tolerancia de 9% => Umbral de 82.09 a 90.50


,question,pct_nan,pct_0.0,pct_2.0,pct_3.0,pct_7.0,pct_8.0,pct_9.0,pct_99.0,limited
4,pni001c,85.494223,4.813864,9.691913,0.0,0.0,0.0,0.0,0.0,True
12,pni003,85.494223,5.006418,9.499358,0.0,0.0,0.0,0.0,0.0,True




------------------------------------------------------------------------------------------------------------------------------------------------------




Al revisar las composiciones es sencillo notar que las siguientes no cuentan como "limited"(*), es decir que sus respuestas no caen en el conjunto de codificaciones categóricas estándar del test (SÍ=2, NO=0, etc.):

| Código de pregunta | Padecimiento | Razón |
|---|---|---|
| pea001z | Bulimia |Corresponde a la codificación de la estatura en sistema métrico|
| pea002k | Bulimia |Corresponde a la codificación del peso actual en kilogramos|
| pea003k | Bulimia |Corresponde a la codificación del menor peso en kilogramos|
| pea004 | Bulimia |Se incluye una codificación anómala "1" como valor categórico, posiblemente "Tal Vez"|


(*): No se consideran las preguntas pal001_, pmj001_, psu003_ o pni001_, aunque aparezcan en el output, pues son contingentes

Además, es importante notar que las preguntas basales correspondientes a manía e hipomanía tienen más de un 99% de valores perdidos. No se considerarán a efectos del LCA

In [8]:
codes = {
    "pso": [(1, 5)],
    "pga": [(1, 5), (23, 29)],
    "psa": [(1, 12)],
    "pea": [(1, 4), (10, 12), (20, 20)],
    "pmd": [(1, 22), (35, 35)],
    "psz": [(1, 2), (9, 14)],
    "pad": [(1, 2), (4, 11), (22, 25), (27, 29), (31, 33), (44, 44)],
    "pcd": [(1, 4), (7, 9), (12, 15), (19, 23), (25, 25), (27, 29), (39, 39)],
    "pod": [(1, 12)],
    "pal": [(1, 1)],
    "pmj": [(1, 1)],
    "psu": [(1, 11)],
    "pni": [(1, 2)]
}

replacements = {
    'pea001': 'pea001z',
    'pea002': 'pea002k',
    'pea003': 'pea003k'
}

# Generar lista original
question_list = [
    f"{code}{num:03d}"
    for code, ranges in codes.items()
    for start, end in ranges
    for num in range(start, end + 1)
]


# **SOLUCIÓN CORRECTA:**
# Opción A: Usando función auxiliar
def apply_replacements(text, rep_dict):
    for old, new in rep_dict.items():
        text = text.replace(old, new)
    return text

question_list = [apply_replacements(item, replacements) for item in question_list]
df_sintomas_disc[question_list]

,pso001,pso002,pso003,pso004,pso005,pga001,pga002,pga003,pga004,pga005,...,psu004,psu005,psu006,psu007,psu008,psu009,psu010,psu011,pni001,pni002
0,2.0,2.0,2.0,0.0,0.0,0.0,0.0,0.0,2.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
1,2.0,2.0,2.0,0.0,2.0,2.0,0.0,2.0,2.0,2.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
2,2.0,0.0,2.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
3,2.0,2.0,0.0,0.0,0.0,2.0,2.0,2.0,2.0,2.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
4,2.0,2.0,0.0,0.0,0.0,2.0,0.0,2.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1553,2.0,0.0,0.0,0.0,2.0,2.0,0.0,0.0,2.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,2.0,0.0
1554,2.0,2.0,0.0,0.0,2.0,0.0,0.0,2.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,2.0,0.0
1555,2.0,2.0,0.0,0.0,2.0,2.0,0.0,2.0,2.0,2.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
1556,2.0,2.0,0.0,0.0,0.0,2.0,0.0,2.0,2.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


In [9]:
# Definir el mapa de códigos global
CODE_MAP = {
    0: "NO", 2: "SI", 3: "AV", 1: "AV",
    7: "NR", 77: "NR",
    8: "NA", 88: "NA", 
    9: "NS", 99: "NS"
}

def calculate_imc_categories(df, weight_col='pea002k', min_weight_col='pea003k', height_col='pea001z'):
    """
    Calcula el IMC basado en peso y estatura, y categoriza en las categorías especificadas.
    
    Parameters:
    -----------
    df : DataFrame
        DataFrame con los datos
    weight_col : str
        Columna del peso actual en kilos (default: 'pea002k')
    min_weight_col : str
        Columna del menor peso en kilos (default: 'pea003k')
    height_col : str
        Columna de la estatura en centímetros (default: 'pea001z')
    
    Returns:
    --------
    DataFrame con dos nuevas columnas IMC_pea002 e IMC_pea003, y sin las columnas originales
    """
    
    df = df.copy()
    
    # Verificar que las columnas existan
    required_cols = [weight_col, min_weight_col, height_col]
    missing_cols = [col for col in required_cols if col not in df.columns]
    
    if missing_cols:
        print(f"Advertencia: Columnas no encontradas: {missing_cols}")
        return df
    
    # Convertir la estatura a metros (está en cm)
    df['height_m'] = df[height_col] / 100
    
    # Calcular IMC para peso actual
    df['IMC_pea002'] = df[weight_col] / (df['height_m'] ** 2)
    
    # Calcular IMC para menor peso
    df['IMC_pea003'] = df[min_weight_col] / (df['height_m'] ** 2)
    
    # Definir categorías del IMC
    imc_categories = ['underweight', 'normal weight', 'overweight', 'obese']
    imc_bins = [0, 18.5, 25, 30, np.inf]
    
    # Categorizar IMC para peso actual
    df['IMC_pea002'] = pd.cut(
        df['IMC_pea002'], 
        bins=imc_bins, 
        labels=imc_categories, 
        right=False
    )
    
    # Categorizar IMC para menor peso
    df['IMC_pea003'] = pd.cut(
        df['IMC_pea003'], 
        bins=imc_bins, 
        labels=imc_categories, 
        right=False
    )
    
    # Eliminar columnas temporales y originales
    df = df.drop(columns=['height_m', weight_col, min_weight_col, height_col])
    
    return df


def apply_code_map(df, columns, code_map=CODE_MAP):
    """
    Aplica el mapeo de códigos a las columnas especificadas.
    
    Parameters:
    -----------
    df : DataFrame
        DataFrame con los datos
    columns : list
        Lista de columnas a mapear
    code_map : dict
        Diccionario de mapeo de códigos
    
    Returns:
    --------
    DataFrame con códigos mapeados
    """
    df = df.copy()
    
    for col in columns:
        if col in df.columns:
            # Aplicar el mapeo, manteniendo los valores que no están en el mapa
            df[col] = df[col].apply(lambda x: code_map.get(x, x) if pd.notnull(x) else x)
    
    return df


def perform_lca_clustering(df, question_list, id_column='id', n_components_range=range(2, 7)):
    """
    Realiza LCA clustering con todas las variables como categóricas.
    """
    warnings.filterwarnings('ignore')
    
    print("="*70)
    print("ANÁLISIS DE CLASES LATENTES (LCA)")
    print("="*70)
    
    # PASO 1: Crear DataFrame de trabajo
    work_df = df.copy()
    print(f"\n[1] Preparando datos: {len(work_df)} observaciones")
    
    # PASO 2: Procesar columnas de IMC (si existen)
    imc_required_cols = ['pea001z', 'pea002k', 'pea003k']
    
    # Verificar si tenemos columnas de IMC en question_list
    imc_cols_in_question_list = [col for col in imc_required_cols if col in question_list]
    
    if imc_cols_in_question_list and all(col in work_df.columns for col in imc_required_cols):
        print(f"[2] Procesando columnas de IMC...")
        print(f"    Transformando peso y estatura a categorías de IMC...")
        
        # Calcular y categorizar IMC
        work_df = calculate_imc_categories(work_df)
        
        # Actualizar question_list: eliminar columnas originales, agregar categorías
        for col in imc_required_cols:
            if col in question_list:
                question_list.remove(col)
        
        # Agregar nuevas columnas de IMC categorizadas
        imc_new_cols = ['IMC_pea002', 'IMC_pea003']
        for new_col in imc_new_cols:
            if new_col in work_df.columns:
                question_list.append(new_col)
        
        print(f"    ✓ IMC categorizado: {imc_new_cols}")
    
    # PASO 3: Aplicar mapeo de códigos
    print(f"\n[3] Aplicando mapeo de códigos...")
    
    # Verificar que las columnas existan
    existing_cols = [col for col in question_list if col in work_df.columns]
    missing_cols = [col for col in question_list if col not in work_df.columns]
    
    if missing_cols:
        print(f"    ⚠ Columnas no encontradas: {missing_cols}")
        question_list = existing_cols
    
    work_df = apply_code_map_without_imputation(work_df, question_list)
    print(f"    ✓ Mapeo aplicado a {len(question_list)} columnas")
    
    # VERIFICACIÓN CRÍTICA: Todas deben ser categóricas
    print(f"\n[4] Verificando tipos de datos...")
    non_categorical = []
    for col in question_list:
        if col in work_df.columns:
            dtype = work_df[col].dtype
            if dtype not in ['object', 'category']:
                non_categorical.append((col, dtype))
    
    if non_categorical:
        print(f"    ⚠ ADVERTENCIA: Columnas no categóricas encontradas:")
        for col, dtype in non_categorical:
            print(f"       {col}: {dtype}")
        print(f"    Estas columnas se forzarán a categóricas para LCA")
    
    # PASO 5: Análisis de valores faltantes
    print(f"\n[5] Análisis de valores faltantes (NaNs):")
    nan_df = work_df[question_list].copy()
    
    for col in nan_df.columns:
        nan_count = nan_df[col].isna().sum()
        if nan_count > 0:
            nan_percent = (nan_count / len(nan_df)) * 100
            print(f"    {col:<15}: {nan_count:>4} NaNs ({nan_percent:>6.2f}%)")
    
    total_nans = nan_df.isna().sum().sum()
    total_cells = nan_df.shape[0] * nan_df.shape[1]
    total_nan_percent = (total_nans / total_cells) * 100
    print(f"\n    Total: {total_nans:,} NaNs en {total_cells:,} celdas ({total_nan_percent:.2f}%)")
    
    # PASO 6: Preparar datos para LCA (todas como categóricas)
    print(f"\n[6] Preparando datos para LCA (todas variables como categóricas)...")
    
    cluster_df = work_df[question_list].copy()
    
    # Forzar todas las columnas a categóricas
    for col in cluster_df.columns:
        # Convertir a string para manejo uniforme
        cluster_df[col] = cluster_df[col].astype(str)
    
    # Codificar variables categóricas para análisis
    le_dict = {}
    encoded_data = []
    feature_names = []
    
    for col in cluster_df.columns:
        le = LabelEncoder()
        
        # Manejar valores NaN: tratarlos como categoría especial
        col_data = cluster_df[col].copy()
        col_data = col_data.fillna('MISSING')  # Marcar NaN explícitamente
        
        # Codificar
        try:
            encoded = le.fit_transform(col_data)
            encoded_data.append(encoded)
            feature_names.append(col)
            le_dict[col] = le
        except Exception as e:
            print(f"    ✗ Error codificando {col}: {e}")
    
    if not encoded_data:
        print("    ✗ No se pudieron codificar datos para LCA")
        return work_df
    
    # Crear matriz de características
    X = np.column_stack(encoded_data)
    print(f"    ✓ Datos preparados: {X.shape[0]} × {X.shape[1]}")
    print(f"    ✓ Columnas codificadas: {len(feature_names)}")
    
    # PASO 7: Realizar LCA
    print(f"\n[7] Realizando Análisis de Clases Latentes...")
    
    results = {}
    metrics = []
    
    for n_components in n_components_range:
        print(f"\n    k = {n_components} clases latentes...")
        
        try:
            # Para variables categóricas, podemos usar LatentClassAnalysis
            # o seguir con GaussianMixture con covarianza diagonal
            gmm = GaussianMixture(
                n_components=n_components,
                covariance_type='diag',  # Mejor para variables categóricas
                random_state=42,
                n_init=10,
                max_iter=500,
                tol=1e-3
            )
            
            clusters = gmm.fit_predict(X)
            
            # Almacenar resultados
            results[n_components] = clusters
            
            # Calcular métricas
            bic = gmm.bic(X)
            aic = gmm.aic(X)
            
            metrics.append({
                'k': n_components,
                'BIC': bic,
                'AIC': aic,
                'Convergió': gmm.converged_,
                'Iteraciones': gmm.n_iter_
            })
            
            print(f"      ✓ BIC: {bic:.2f}, AIC: {aic:.2f}")
            
            # Distribución de clusters
            unique, counts = np.unique(clusters, return_counts=True)
            print(f"      ✓ Distribución:")
            for cluster, count in sorted(zip(unique, counts)):
                percentage = (count / len(clusters)) * 100
                print(f"          Clase {cluster}: {count} ({percentage:.1f}%)")
            
        except Exception as e:
            print(f"      ✗ Error: {str(e)}")
            results[n_components] = None
    
    # PASO 8: Crear DataFrame de salida
    print(f"\n[8] Creando DataFrame de salida...")
    
    # DataFrame base con id y columnas originales (mapeadas)
    output_cols = [id_column] + question_list
    available_cols = [col for col in output_cols if col in work_df.columns]
    output_df = work_df[available_cols].copy()
    
    # Agregar columnas de clusters
    for n_components, clusters in results.items():
        if clusters is not None:
            col_name = f'lca_k{n_components}'
            output_df[col_name] = clusters
            print(f"     Añadida columna: {col_name}")
    
    # PASO 9: Resumen final
    print(f"\n" + "="*70)
    print("RESUMEN FINAL - LCA CON VARIABLES CATEGÓRICAS")
    print("="*70)
    
    print(f"\nDataFrame de salida:")
    print(f"  • Dimensiones: {output_df.shape[0]} filas × {output_df.shape[1]} columnas")
    print(f"  • Variables categóricas: {len(question_list)}")
    
    if metrics:
        metrics_df = pd.DataFrame(metrics)
        print(f"\nMétricas de modelos:")
        print(metrics_df.to_string(index=False))
        
        # Recomendar mejor modelo (menor BIC)
        valid_metrics = metrics_df[metrics_df['BIC'].notnull()]
        if not valid_metrics.empty:
            best_model = valid_metrics.loc[valid_metrics['BIC'].idxmin()]
            print(f"\n  → Modelo recomendado: k = {best_model['k']} (menor BIC)")
    
    print(f"\n" + "="*70)
    print("PROCESO COMPLETADO")
    print("="*70)
    
    return output_df

def apply_code_map_without_imputation(df, columns, code_map=CODE_MAP):
    """
    Aplica el mapeo de códigos a las columnas especificadas sin imputar valores NaN.
    """
    df_result = df.copy()
    
    for col in columns:
        if col in df_result.columns:
            # Solo aplicar el mapeo a valores no nulos
            df_result[col] = df_result[col].apply(
                lambda x: code_map.get(x, x) if pd.notnull(x) else x
            )
    
    return df_result

df=df_sintomas_disc.copy()
final_df=perform_lca_clustering(df, question_list, id_column='id')

ANÁLISIS DE CLASES LATENTES (LCA)

[1] Preparando datos: 1558 observaciones
[2] Procesando columnas de IMC...
    Transformando peso y estatura a categorías de IMC...
    ✓ IMC categorizado: ['IMC_pea002', 'IMC_pea003']

[3] Aplicando mapeo de códigos...
    ✓ Mapeo aplicado a 136 columnas

[4] Verificando tipos de datos...

[5] Análisis de valores faltantes (NaNs):
    pso001         :    1 NaNs (  0.06%)
    pso002         :    1 NaNs (  0.06%)
    pso003         :    1 NaNs (  0.06%)
    pso004         :    1 NaNs (  0.06%)
    pso005         :    1 NaNs (  0.06%)
    pga001         :   45 NaNs (  2.89%)
    pga002         :    2 NaNs (  0.13%)
    pga003         :    2 NaNs (  0.13%)
    pga004         :    2 NaNs (  0.13%)
    pga005         :    2 NaNs (  0.13%)
    pga023         :    2 NaNs (  0.13%)
    pga024         :    2 NaNs (  0.13%)
    pga025         :    2 NaNs (  0.13%)
    pga026         :    2 NaNs (  0.13%)
    pga027         :    2 NaNs (  0.13%)
    pga028      

In [10]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.stats import chi2_contingency
from itertools import combinations

print("=" * 100)
print("ANÁLISIS ESTADÍSTICO ")
print("=" * 100)

# Configuración para visualización completa
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)
np.set_printoptions(precision=3, suppress=True)

# Identificar columnas de cluster
cluster_cols = sorted([col for col in final_df.columns if 'lca_k' in col])
print(f"\nCOLUMNAS DE CLUSTER ENCONTRADAS: {cluster_cols}")
print(f"Total de observaciones: {len(final_df)}")

# Identificar TODAS las variables categóricas (excluyendo columnas de cluster, ID y IMC)
cat_vars = [col for col in final_df.columns 
            if col not in cluster_cols 
            and col != 'id' 
            and not col.startswith('IMC_')
            and final_df[col].dtype in ['object', 'category', 'string']]

print(f"TOTAL variables categóricas identificadas: {len(cat_vars)}")

# ============================================================================
# ANÁLISIS 1: DISTRIBUCIÓN DE CLUSTERS Y ESTADÍSTICAS BÁSICAS
# ============================================================================
print("\n" + "=" * 80)
print("1. ESTADÍSTICAS DE DISTRIBUCIÓN DE CLUSTERS POR CLUSTERIZACIÓN")
print("=" * 80)

# Crear DataFrame de resumen
resumen_clusters = pd.DataFrame()

for cluster_col in cluster_cols:
    cluster_stats = final_df[cluster_col].value_counts().sort_index()
    n_clusters = len(cluster_stats)
    
    # Calcular estadísticas detalladas
    for cluster_id in cluster_stats.index:
        cluster_data = final_df[final_df[cluster_col] == cluster_id]
        cluster_size = len(cluster_data)
        
        # Calcular estadísticas de completitud para este cluster
        missing_stats = []
        for var in cat_vars:
            if var in cluster_data.columns:
                missing_pct = cluster_data[var].isna().mean() * 100
                if missing_pct < 50:  # Solo considerar variables con menos del 50% missing
                    missing_stats.append(missing_pct)
        
        avg_missing = np.mean(missing_stats) if missing_stats else 100
        
        resumen_clusters = pd.concat([
            resumen_clusters,
            pd.DataFrame([{
                'Clusterizacion': cluster_col,
                'Cluster': cluster_id,
                'N_Observaciones': cluster_size,
                'Porcentaje_Total': (cluster_size / len(final_df)) * 100,
                'Missing_Promedio': avg_missing,
                'Completitud_Promedio': 100 - avg_missing
            }])
        ])

# Mostrar resumen consolidado
print("\nRESUMEN DE TODOS LOS CLUSTERS:")
for cluster_col in cluster_cols:
    cluster_data = resumen_clusters[resumen_clusters['Clusterizacion'] == cluster_col]
    total_obs = cluster_data['N_Observaciones'].sum()
    
    print(f"\n{cluster_col}:")
    print(f"  Número de clusters: {len(cluster_data)}")
    print(f"  Total observaciones: {total_obs}")
    print(f"  Tamaño promedio de cluster: {cluster_data['N_Observaciones'].mean():.0f}")
    print(f"  Rango de tamaños: {cluster_data['N_Observaciones'].min():.0f} - {cluster_data['N_Observaciones'].max():.0f}")
    print(f"  Completitud promedio: {cluster_data['Completitud_Promedio'].mean():.1f}%")

# ============================================================================
# ANÁLISIS 2: IDENTIFICACIÓN DE VARIABLES CLAVE POR CLUSTER
# ============================================================================
print("\n" + "=" * 80)
print("2. IDENTIFICACIÓN SISTEMÁTICA DE VARIABLES CLAVE")
print("=" * 80)

# Para cada clusterización, identificar las variables más discriminantes
variables_discriminantes_global = {}

for cluster_col in cluster_cols:
    print(f"\nAnalizando variables para {cluster_col}...")
    
    discriminacion_scores = []
    
    # Analizar TODAS las variables categóricas
    for var in cat_vars:
        if var in final_df.columns:
            try:
                # Crear tabla de contingencia
                contingency_table = pd.crosstab(final_df[var], final_df[cluster_col])
                
                # Calcular chi-cuadrado y Cramér's V
                chi2, p, dof, expected = chi2_contingency(contingency_table)
                n = contingency_table.sum().sum()
                min_dim = min(contingency_table.shape) - 1
                cramers_v = np.sqrt(chi2 / (n * min_dim)) if min_dim > 0 else 0
                
                # Calcular poder discriminante adicional
                discriminacion_total = 0
                for cluster_id in contingency_table.columns:
                    cluster_vals = contingency_table[cluster_id]
                    total_cluster = cluster_vals.sum()
                    
                    if total_cluster > 0:
                        # Calcular concentración en categorías principales
                        top_2_cats = cluster_vals.nlargest(2)
                        concentracion = top_2_cats.sum() / total_cluster
                        discriminacion_total += concentracion
                
                discriminacion_promedio = discriminacion_total / len(contingency_table.columns)
                
                discriminacion_scores.append({
                    'Variable': var,
                    'Cramers_V': cramers_v,
                    'p_value': p,
                    'Chi2': chi2,
                    'Discriminacion_Promedio': discriminacion_promedio,
                    'Significativo': p < 0.001
                })
                
            except Exception as e:
                continue
    
    # Convertir a DataFrame y ordenar
    discriminacion_df = pd.DataFrame(discriminacion_scores)
    
    # Filtrar solo variables significativas
    variables_significativas = discriminacion_df[discriminacion_df['Significativo']].copy()
    
    # Ordenar por Cramér's V descendente
    variables_significativas = variables_significativas.sort_values('Cramers_V', ascending=False)
    
    print(f"  Variables significativas encontradas: {len(variables_significativas)} de {len(cat_vars)}")
    
    if len(variables_significativas) > 0:
        # Guardar para uso posterior
        variables_discriminantes_global[cluster_col] = variables_significativas
        
        # Mostrar top 15 variables más discriminantes
        print(f"\n  Top 25 variables más discriminantes para {cluster_col}:")
        print("  " + "-" * 90)
        print("  Variable                                Cramér's V   p-value    χ²       Discriminación")
        print("  " + "-" * 90)
        
        for idx, row in variables_significativas.head(25).iterrows():
            print(f"  {row['Variable'][:40]:40} {row['Cramers_V']:.4f}     {row['p_value']:.2e}  {row['Chi2']:.0f}       {row['Discriminacion_Promedio']:.3f}")

# ============================================================================
# ANÁLISIS 3: PERFIL COMPLETO POR CLUSTER - VERSIÓN CON FRECUENCIAS ABSOLUTAS
# ============================================================================
print("\n" + "=" * 80)
print("3. PERFIL COMPLETO POR CLUSTER (ANÁLISIS CON 25 VARIABLES MÁS DISCRIMINANTES)")
print("=" * 80)

for cluster_col in cluster_cols:
    print(f"\n{'#' * 70}")
    print(f"PERFILES DETALLADOS PARA: {cluster_col}")
    print(f"{'#' * 70}")
    
    if cluster_col in variables_discriminantes_global:
        top_variables = variables_discriminantes_global[cluster_col].head(25)['Variable'].tolist()
    else:
        top_variables = cat_vars
    
    print(f"Analizando {len(top_variables)} variables clave...")
    
    # Tamaño total del dataset para cálculos globales
    total_dataset = len(final_df)
    
    perfiles_detallados = []
    
    for cluster_id in sorted(final_df[cluster_col].unique()):
        cluster_data = final_df[final_df[cluster_col] == cluster_id]
        cluster_size = len(cluster_data)
        
        perfil_cluster = {
            'Cluster': cluster_id,
            'Tamaño': cluster_size,
            'Porcentaje': (cluster_size / total_dataset) * 100,
            'Variables_Destacadas': []
        }
        
        for var in top_variables:
            if var in cluster_data.columns:
                # FRECUENCIAS ABSOLUTAS en el cluster
                cluster_counts = cluster_data[var].value_counts()
                # FRECUENCIAS RELATIVAS en el cluster
                cluster_dist = cluster_counts / cluster_size
                
                # FRECUENCIAS ABSOLUTAS globales
                global_counts = final_df[var].value_counts()
                # FRECUENCIAS RELATIVAS globales
                global_dist = global_counts / total_dataset
                
                # Filtrar categorías con al menos 5% en el cluster
                cluster_dist_filtrado = cluster_dist[cluster_dist >= 0.05]
                
                if len(cluster_dist_filtrado) == 0:
                    cluster_dist_filtrado = cluster_dist.head(3)
                
                caracteristicas_sobre = []
                caracteristicas_sub = []
                
                for cat, freq_cluster_rel in cluster_dist_filtrado.items():
                    freq_global_rel = global_dist.get(cat, 0)
                    
                    # Obtener conteos absolutos
                    count_cluster = cluster_counts.get(cat, 0)
                    count_global = global_counts.get(cat, 0)
                    
                    # Calcular diferencia
                    diferencia = freq_cluster_rel - freq_global_rel
                    
                    # Si la diferencia es significativamente positiva (>10%)
                    if diferencia > 0.10:
                        caracteristicas_sobre.append({
                            'Categoria': cat,
                            'Freq_Cluster_Rel': freq_cluster_rel,
                            'Count_Cluster': count_cluster,
                            'Freq_Global_Rel': freq_global_rel,
                            'Count_Global': count_global,
                            'Diferencia': diferencia,
                            'Tipo': 'SOBRE'
                        })
                    # Si la diferencia es significativamente negativa (< -10%)
                    elif diferencia < -0.10:
                        caracteristicas_sub.append({
                            'Categoria': cat,
                            'Freq_Cluster_Rel': freq_cluster_rel,
                            'Count_Cluster': count_cluster,
                            'Freq_Global_Rel': freq_global_rel,
                            'Count_Global': count_global,
                            'Diferencia': diferencia,
                            'Tipo': 'SUB'
                        })
                
                # Ordenar por valor absoluto de diferencia
                caracteristicas_sobre.sort(key=lambda x: abs(x['Diferencia']), reverse=True)
                caracteristicas_sub.sort(key=lambda x: abs(x['Diferencia']), reverse=True)
                
                todas_caracteristicas = caracteristicas_sobre + caracteristicas_sub
                
                if todas_caracteristicas:
                    max_dif_abs = max([abs(c['Diferencia']) for c in todas_caracteristicas], default=0)
                    
                    perfil_cluster['Variables_Destacadas'].append({
                        'Variable': var,
                        'Caracteristicas_Sobre': caracteristicas_sobre,
                        'Caracteristicas_Sub': caracteristicas_sub,
                        'Todas_Caracteristicas': todas_caracteristicas,
                        'Num_Sobre': len(caracteristicas_sobre),
                        'Num_Sub': len(caracteristicas_sub),
                        'Max_Diferencia_Abs': max_dif_abs
                    })
        
        # Ordenar variables por máxima diferencia absoluta
        perfil_cluster['Variables_Destacadas'].sort(
            key=lambda x: x['Max_Diferencia_Abs'],
            reverse=True
        )
        
        perfiles_detallados.append(perfil_cluster)
    
    # Mostrar resultados con AMBAS frecuencias (relativas y absolutas)
    for perfil in perfiles_detallados:
        print(f"\n  CLUSTER {perfil['Cluster']}: {perfil['Tamaño']} observaciones ({perfil['Porcentaje']:.1f}%)")
        print(f"  {'─' * 80}")
        
        if perfil['Variables_Destacadas']:
            print(f"  Variables distintivas:")
            
            for idx, var_info in enumerate(perfil['Variables_Destacadas'], 1):
                var_name = var_info['Variable']
                caracteristicas_sobre = var_info['Caracteristicas_Sobre']
                caracteristicas_sub = var_info['Caracteristicas_Sub']
                total_caracteristicas = len(caracteristicas_sobre) + len(caracteristicas_sub)
                
                if total_caracteristicas > 0:
                    print(f"\n    {idx:2d}. {var_name}:")
                    
                    # Mostrar categorías sobrerrepresentadas
                    if caracteristicas_sobre:
                        for i, carac in enumerate(caracteristicas_sobre, 1):
                            print(f"        SOBRE {i}: '{carac['Categoria']}' - "
                                  f"{carac['Freq_Cluster_Rel']:.0%} ({carac['Count_Cluster']}/{perfil['Tamaño']}) "
                                  f"vs {carac['Freq_Global_Rel']:.0%} ({carac['Count_Global']}/{total_dataset}) global, "
                                  f"+{abs(carac['Diferencia'])*100:.0f}pp")
                    
                    # Mostrar categorías subrepresentadas
                    if caracteristicas_sub:
                        for i, carac in enumerate(caracteristicas_sub, 1):
                            print(f"        SUB {i}: '{carac['Categoria']}' - "
                                  f"{carac['Freq_Cluster_Rel']:.0%} ({carac['Count_Cluster']}/{perfil['Tamaño']}) "
                                  f"vs {carac['Freq_Global_Rel']:.0%} ({carac['Count_Global']}/{total_dataset}) global, "
                                  f"-{abs(carac['Diferencia'])*100:.0f}pp")
        else:
            print(f"  No se encontraron variables distintivas (diferencia >10% con global)")
        
        print(f"  {'─' * 80}")

# ============================================================================
# ANÁLISIS 5: PATRONES COMBINADOS E INTERACCIONES ENTRE 25 VARIABLES CLAVE
# ============================================================================
print("\n" + "=" * 80)
print("5. ANÁLISIS DE PATRONES COMBINADOS E INTERACCIONES ENTRE 25 VARIABLES CLAVE")
print("=" * 80)

# Para cada clusterización, analizar combinaciones de variables
for cluster_col in cluster_cols:
    print(f"\n{'#' * 70}")
    print(f"ANÁLISIS DE PATRONES COMBINADOS PARA: {cluster_col}")
    print(f"{'#' * 70}")
    
    # Obtener las 25 variables más discriminantes
    if cluster_col in variables_discriminantes_global:
        top_vars = variables_discriminantes_global[cluster_col].head(25)['Variable'].tolist()
        # Mostrar las variables que se van a analizar
        print(f"\n  Variables analizadas (25 más discriminantes):")
        for i in range(0, 25, 5):
            grupo = top_vars[i:min(i+5, 25)]
            print(f"    {i+1:2d}-{min(i+5, 25):2d}: {', '.join(grupo)}")
    else:
        top_vars = cat_vars[:25]
    
    # Analizar combinaciones de 2 variables (pares)
    patrones_combinados = []
    n_clusters = final_df[cluster_col].nunique()
    
    print(f"\n  Analizando combinaciones entre las 25 variables...")
    print(f"  Total de combinaciones posibles: {(25*24)//2} pares")
    print(f"  Se evaluarán {n_clusters} clusters")
    
    # Progreso
    total_combinaciones = 0
    combinaciones_analizadas = 0
    
    # Calcular Cramér's V global para cada par de variables primero (optimización)
    print(f"  Calculando asociaciones globales...")
    cramers_global = {}
    
    for i in range(len(top_vars)):
        for j in range(i+1, len(top_vars)):
            var1, var2 = top_vars[i], top_vars[j]
            if var1 in final_df.columns and var2 in final_df.columns:
                try:
                    cross_table = pd.crosstab(final_df[var1], final_df[var2])
                    chi2, p, dof, expected = chi2_contingency(cross_table)
                    n = cross_table.sum().sum()
                    min_dim = min(cross_table.shape) - 1
                    cramers_v = np.sqrt(chi2 / (n * min_dim)) if min_dim > 0 else 0
                    cramers_global[(var1, var2)] = cramers_v
                except:
                    cramers_global[(var1, var2)] = 0
    
    print(f"  Analizando patrones por cluster...")
    
    # Para cada cluster, analizar las asociaciones
    for cluster_id in sorted(final_df[cluster_col].unique()):
        cluster_data = final_df[final_df[cluster_col] == cluster_id]
        cluster_size = len(cluster_data)
        
        # Solo analizar clusters con tamaño suficiente
        if cluster_size < 20:
            continue
            
        print(f"    Cluster {cluster_id} ({cluster_size} observaciones)...")
        
        # Analizar cada par de variables
        for i in range(len(top_vars)):
            for j in range(i+1, len(top_vars)):
                var1, var2 = top_vars[i], top_vars[j]
                
                if var1 not in cluster_data.columns or var2 not in cluster_data.columns:
                    continue
                
                try:
                    # Calcular Cramér's V para este cluster
                    cluster_cross = pd.crosstab(cluster_data[var1], cluster_data[var2])
                    
                    if cluster_cross.size == 0:
                        continue
                    
                    chi2_cluster, p_cluster, dof_cluster, expected_cluster = chi2_contingency(cluster_cross)
                    n_cluster = cluster_cross.sum().sum()
                    min_dim_cluster = min(cluster_cross.shape) - 1
                    
                    if min_dim_cluster == 0:
                        continue
                        
                    cramers_v_cluster = np.sqrt(chi2_cluster / (n_cluster * min_dim_cluster))
                    
                    # Obtener Cramér's V global para este par
                    cramers_v_global = cramers_global.get((var1, var2), 0)
                    
                    # Calcular diferencia absoluta
                    diferencia_abs = abs(cramers_v_cluster - cramers_v_global)
                    
                    # Solo considerar diferencias significativas
                    if diferencia_abs > 0.15:  # Umbral reducido para capturar más patrones
                        # Calcular si es más fuerte o más débil
                        if cramers_v_cluster > cramers_v_global:
                            tipo = "MÁS FUERTE"
                        else:
                            tipo = "MÁS DÉBIL"
                        
                        # Calcular fuerza de la asociación en el cluster
                        if cramers_v_cluster >= 0.5:
                            fuerza_cluster = "FUERTE"
                        elif cramers_v_cluster >= 0.3:
                            fuerza_cluster = "MODERADA"
                        elif cramers_v_cluster >= 0.1:
                            fuerza_cluster = "DÉBIL"
                        else:
                            fuerza_cluster = "MUY DÉBIL"
                        
                        # Almacenar patrón
                        patrones_combinados.append({
                            'Variables': f"{var1} × {var2}",
                            'Cluster': cluster_id,
                            'Tamaño_Cluster': cluster_size,
                            'Asociacion_Global': cramers_v_global,
                            'Asociacion_Cluster': cramers_v_cluster,
                            'Diferencia': cramers_v_cluster - cramers_v_global,
                            'Diferencia_Abs': diferencia_abs,
                            'Tipo_Cambio': tipo,
                            'Fuerza_Cluster': fuerza_cluster,
                            'p_valor': p_cluster
                        })
                        
                except Exception as e:
                    # Silenciosamente continuar con el siguiente par
                    continue
    
    # Mostrar resultados
    if not patrones_combinados:
        print(f"\n  No se encontraron patrones de interacción significativos.")
        print(f"  (Ninguna combinación mostró diferencia > 0.15 en Cramér's V)")
        continue
    
    print(f"\n  RESUMEN: Se encontraron {len(patrones_combinados)} patrones de interacción significativos")
    print(f"  (Diferencia > 0.15 en coeficiente Cramér's V)")
    
    # Agrupar por cluster para mejor presentación
    clusters_unicos = sorted(set([p['Cluster'] for p in patrones_combinados]))
    
    for cluster_id in clusters_unicos:
        patrones_cluster = [p for p in patrones_combinados if p['Cluster'] == cluster_id]
        
        # Ordenar por diferencia absoluta descendente
        patrones_cluster.sort(key=lambda x: x['Diferencia_Abs'], reverse=True)
        
        # Mostrar top 10 para este cluster
        top_n = min(10, len(patrones_cluster))
        
        print(f"\n  --- CLUSTER {cluster_id} ---")
        print(f"  {len(patrones_cluster)} patrones de interacción significativos encontrados")
        print(f"  (Mostrando los {top_n} más fuertes)")
        print(f"  {'─' * 60}")
        
        for idx, patron in enumerate(patrones_cluster[:top_n], 1):
            # Formatear nombres de variables (cortar si son muy largos)
            var_names = patron['Variables']
            if len(var_names) > 50:
                var_names = var_names[:47] + "..."
            
            # Determinar símbolo según tipo de cambio
            if patron['Tipo_Cambio'] == "MÁS FUERTE":
                simbolo = "↑"
            else:
                simbolo = "↓"
            
            print(f"  {idx:2d}. {simbolo} {var_names}")
            print(f"       Asociación: Cluster={patron['Asociacion_Cluster']:.3f} ({patron['Fuerza_Cluster']}) "
                  f"vs Global={patron['Asociacion_Global']:.3f}")
            print(f"       Diferencia: {patron['Diferencia']:+.3f} ({patron['Tipo_Cambio']})")
            print(f"       p-valor: {patron['p_valor']:.4f}")
            print()
    
    # Análisis agregado por tipo de cambio
    print(f"\n  ANÁLISIS AGREGADO PARA {cluster_col}:")
    print(f"  {'─' * 60}")
    
    # Contar patrones por tipo de cambio
    patrones_fuertes = [p for p in patrones_combinados if p['Tipo_Cambio'] == "MÁS FUERTE"]
    patrones_debiles = [p for p in patrones_combinados if p['Tipo_Cambio'] == "MÁS DÉBIL"]
    
    print(f"  Patrones con asociación MÁS FUERTE en algún cluster: {len(patrones_fuertes)}")
    print(f"  Patrones con asociación MÁS DÉBIL en algún cluster: {len(patrones_debiles)}")
    
    # Top 5 patrones con mayor diferencia (positiva y negativa)
    if patrones_combinados:
        print(f"\n  TOP 5 PATRONES CON MAYOR CAMBIO POSITIVO (más fuertes en cluster):")
        patrones_pos = sorted([p for p in patrones_combinados if p['Diferencia'] > 0], 
                             key=lambda x: x['Diferencia'], reverse=True)[:5]
        for idx, patron in enumerate(patrones_pos, 1):
            print(f"    {idx}. Cluster {patron['Cluster']}: {patron['Variables']}")
            print(f"       De {patron['Asociacion_Global']:.3f} a {patron['Asociacion_Cluster']:.3f} "
                  f"(Δ={patron['Diferencia']:.3f})")
        
        print(f"\n  TOP 5 PATRONES CON MAYOR CAMBIO NEGATIVO (más débiles en cluster):")
        patrones_neg = sorted([p for p in patrones_combinados if p['Diferencia'] < 0], 
                             key=lambda x: x['Diferencia'])[:5]
        for idx, patron in enumerate(patrones_neg, 1):
            print(f"    {idx}. Cluster {patron['Cluster']}: {patron['Variables']}")
            print(f"       De {patron['Asociacion_Global']:.3f} a {patron['Asociacion_Cluster']:.3f} "
                  f"(Δ={patron['Diferencia']:.3f})")
    
    # Análisis por fuerza de asociación en el cluster
    print(f"\n  DISTRIBUCIÓN POR FUERZA DE ASOCIACIÓN EN LOS CLUSTERS:")
    fuerzas = {}
    for patron in patrones_combinados:
        fuerza = patron['Fuerza_Cluster']
        if fuerza not in fuerzas:
            fuerzas[fuerza] = 0
        fuerzas[fuerza] += 1
    
    for fuerza, count in sorted(fuerzas.items(), key=lambda x: x[1], reverse=True):
        porcentaje = (count / len(patrones_combinados)) * 100
        print(f"    {fuerza}: {count} patrones ({porcentaje:.1f}%)")

print("\n" + "=" * 80)
print("FIN DEL ANÁLISIS DE PATRONES COMBINADOS")
print("=" * 80)
# ============================================================================
# ANÁLISIS 6: JERARQUÍA Y ESTABILIDAD DE CLUSTERS
# ============================================================================
print("\n" + "=" * 80)
print("6. ANÁLISIS DE JERARQUÍA Y ESTABILIDAD ENTRE CLUSTERIZACIONES")
print("=" * 80)

# Analizar cómo se relacionan los clusters entre diferentes niveles
if len(cluster_cols) > 1:
    print("\nEvolución jerárquica de clusters:")
    
    # Comparar cada par de clusterizaciones consecutivas
    for i in range(len(cluster_cols) - 1):
        col_actual = cluster_cols[i]
        col_siguiente = cluster_cols[i + 1]
        
        print(f"\n  {col_actual} → {col_siguiente}:")
        print("  " + "-" * 60)
        
        # Calcular matriz de transición en frecuencias absolutas
        tabla_conteo = pd.crosstab(final_df[col_actual], final_df[col_siguiente])
        
        # Calcular matriz de transición en porcentajes
        transicion_porcentaje = pd.crosstab(final_df[col_actual], 
                                           final_df[col_siguiente], 
                                           normalize='index')
        
        # Para cada cluster actual
        for cluster_actual in sorted(final_df[col_actual].unique()):
            # Obtener total de observaciones en este cluster
            total_obs = tabla_conteo.loc[cluster_actual].sum()
            
            # Obtener distribución de porcentajes para este cluster
            distribucion = transicion_porcentaje.loc[cluster_actual]
            
            # Ordenar por porcentaje descendente
            distribucion_ordenada = distribucion.sort_values(ascending=False)
            
            # Filtrar solo destinos con porcentaje > 0 (evitar mostrar ceros)
            distribucion_filtrada = distribucion_ordenada[distribucion_ordenada > 0]
            
            # Construir lista de transiciones
            transiciones = []
            total_acumulado = 0
            
            for cluster_dest, porcentaje in distribucion_filtrada.items():
                # Calcular frecuencia absoluta
                frecuencia = tabla_conteo.loc[cluster_actual, cluster_dest]
                
                # Formatear la transición
                transicion_str = f"{porcentaje:.1%} → {cluster_dest} ({frecuencia} obs)"
                transiciones.append(transicion_str)
                
                # Acumular porcentaje
                total_acumulado += porcentaje
                
                # Mostrar hasta cubrir 99.5% o máximo 5 transiciones
                if total_acumulado >= 0.995 or len(transiciones) >= 5:
                    break
            
            # Verificar si falta alguna parte de la distribución
            if total_acumulado < 0.995 and len(distribucion_filtrada) > len(transiciones):
                # Calcular el porcentaje restante
                porcentaje_restante = 1 - total_acumulado
                
                # Sumar las frecuencias restantes
                clusters_restantes = distribucion_filtrada.index[len(transiciones):]
                frecuencia_restante = sum([tabla_conteo.loc[cluster_actual, dest] 
                                          for dest in clusters_restantes])
                
                if porcentaje_restante > 0.001:  # Solo mostrar si > 0.1%
                    transiciones.append(f"{porcentaje_restante:.1%} → otros ({frecuencia_restante} obs)")
            
            # Imprimir la línea para este cluster
            print(f"  Cluster {cluster_actual} (total={total_obs} obs): {', '.join(transiciones)}")
        
        print("  " + "-" * 60)
        
        # Mostrar también una tabla resumen compacta
        print("\n  Resumen tabular:")
        print("  Clúster | Total obs | Distribución principal")
        print("  " + "-" * 50)
        
        for cluster_actual in sorted(final_df[col_actual].unique()):
            total_obs = tabla_conteo.loc[cluster_actual].sum()
            distribucion = transicion_porcentaje.loc[cluster_actual]
            
            # Obtener las 2 principales transiciones
            principales = distribucion.nlargest(2)
            
            if len(principales) == 2:
                dest1, pct1 = principales.index[0], principales.iloc[0]
                dest2, pct2 = principales.index[1], principales.iloc[1]
                obs1 = tabla_conteo.loc[cluster_actual, dest1]
                obs2 = tabla_conteo.loc[cluster_actual, dest2]
                
                print(f"  {cluster_actual:^7} | {total_obs:^9} | {pct1:.1%}→{dest1}({obs1}), {pct2:.1%}→{dest2}({obs2})")
            elif len(principales) == 1:
                dest, pct = principales.index[0], principales.iloc[0]
                obs = tabla_conteo.loc[cluster_actual, dest]
                print(f"  {cluster_actual:^7} | {total_obs:^9} | {pct:.1%}→{dest}({obs})")
# ============================================================================
# ANÁLISIS 7: SÍNTESIS Y RECOMENDACIONES FINALES
# ============================================================================
print("\n" + "=" * 80)
print("7. SÍNTESIS INTEGRAL")
print("=" * 80)

# Crear resumen evaluativo para cada clusterización
evaluacion_final = []

for cluster_col in cluster_cols:
    n_clusters = final_df[cluster_col].nunique()
    cluster_sizes = final_df[cluster_col].value_counts().values
    
    # Calcular métricas de calidad
    # 1. Balance de clusters
    size_std = cluster_sizes.std()
    size_mean = cluster_sizes.mean()
    balance_score = 1 - (size_std / size_mean) if size_mean > 0 else 0
    
    # 2. Poder discriminante (usando Cramér's V promedio de variables top)
    if cluster_col in variables_discriminantes_global:
        discriminacion_scores = variables_discriminantes_global[cluster_col]['Cramers_V'].head(20)
        discriminacion_score = discriminacion_scores.mean() if len(discriminacion_scores) > 0 else 0
    else:
        discriminacion_score = 0
    
    # 3. Estabilidad (coherencia interna)
    coherencia_scores = []
    for cluster_id in final_df[cluster_col].unique():
        cluster_data = final_df[final_df[cluster_col] == cluster_id]
        
        # Calcular coherencia para variables clave
        for var in cat_vars[:10]:
            if var in cluster_data.columns:
                frecuencias = cluster_data[var].value_counts(normalize=True)
                if not frecuencias.empty:
                    max_freq = frecuencias.iloc[0]
                    coherencia_scores.append(max_freq)
    
    coherencia_score = np.mean(coherencia_scores) if coherencia_scores else 0
    
    # 4. Puntuación combinada
    puntuacion_combinada = (balance_score * 1/3 + 
                           discriminacion_score * 1/3 + 
                           coherencia_score * 1/3)
    
    # Determinar recomendación
    if puntuacion_combinada > 0.7:
        recomendacion = "EXCELENTE"
    elif puntuacion_combinada > 0.5:
        recomendacion = "BUENA"
    elif puntuacion_combinada > 0.3:
        recomendacion = "ACEPTABLE"
    else:
        recomendacion = "LIMITADA"
    
    evaluacion_final.append({
        'Clusterizacion': cluster_col,
        'N_Clusters': n_clusters,
        'Balance': f"{balance_score:.3f}",
        'Discriminacion': f"{discriminacion_score:.3f}",
        'Coherencia': f"{coherencia_score:.3f}",
        'Puntuacion_Total': f"{puntuacion_combinada:.3f}",
        'Recomendacion': recomendacion
    })

# Crear DataFrame de evaluación
eval_df = pd.DataFrame(evaluacion_final)
print("\nEVALUACIÓN COMPARATIVA DE TODAS LAS CLUSTERIZACIONES:")
print("-" * 90)
print(eval_df.sort_values('Puntuacion_Total', ascending=False).to_string(index=False))
print("-" * 90)


ANÁLISIS ESTADÍSTICO 

COLUMNAS DE CLUSTER ENCONTRADAS: ['lca_k2', 'lca_k3', 'lca_k4', 'lca_k5', 'lca_k6']
Total de observaciones: 1558
TOTAL variables categóricas identificadas: 134

1. ESTADÍSTICAS DE DISTRIBUCIÓN DE CLUSTERS POR CLUSTERIZACIÓN

RESUMEN DE TODOS LOS CLUSTERS:

lca_k2:
  Número de clusters: 2
  Total observaciones: 1558
  Tamaño promedio de cluster: 779
  Rango de tamaños: 141 - 1417
  Completitud promedio: 99.2%

lca_k3:
  Número de clusters: 3
  Total observaciones: 1558
  Tamaño promedio de cluster: 519
  Rango de tamaños: 120 - 968
  Completitud promedio: 99.4%

lca_k4:
  Número de clusters: 4
  Total observaciones: 1558
  Tamaño promedio de cluster: 390
  Rango de tamaños: 35 - 918
  Completitud promedio: 98.6%

lca_k5:
  Número de clusters: 5
  Total observaciones: 1558
  Tamaño promedio de cluster: 312
  Rango de tamaños: 36 - 924
  Completitud promedio: 99.1%

lca_k6:
  Número de clusters: 6
  Total observaciones: 1558
  Tamaño promedio de cluster: 260
  Rango

# **Análisis Estadístico de Resultados de Clustering LCA**

##  Información General del Dataset**

| Característica | Valor |
|----------------|-------|
| **Observaciones totales** | 1,558 individuos |
| **Variables categóricas analizadas** | 147 variables |
| **Niveles de clustering** | 5 (lca_k2 a lca_k6) |
| **Completitud de datos** | 95.8% - 99.6% |

## Definiciones de Estadísticas Utilizadas**

###  Cramér's V**
**Fórmula**: 

$V = \sqrt{\frac{\chi^2}{n \times \min(r-1, c-1)}}$


**Interpretación**:
- **0.00-0.10**: Asociación nula/muy débil
- **0.10-0.30**: Asociación débil
- **0.30-0.50**: Asociación moderada
- **0.50-0.70**: Asociación fuerte
- **0.70-1.00**: Asociación muy fuerte/perfecta

**Uso**: Mide qué tan fuertemente cada variable discrimina entre clusters.

### Valor-p (p-value)**
**Interpretación**:
- **< 0.001**: Evidencia estadística muy fuerte
- **< 0.01**: Evidencia fuerte
- **< 0.05**: Evidencia moderada
- **≥ 0.05**: Evidencia insuficiente

**Uso**: Valida significancia estadística de asociaciones.

### Estadístico Chi-cuadrado (χ²)**
**Fórmula**:

$\chi^2 = \sum \frac{(O_{ij} - E_{ij})^2}{E_{ij}}$

Donde $O_{ij}$ = frecuencias observadas, $E_{ij}$ = frecuencias esperadas.

**Uso**: Base para calcular Cramér's V.


###  Métricas de Calidad de Clustering**

| Métrica | Fórmula | Rango | Interpretación |
|---------|---------|-------|----------------|
| **Balance** | $1 - \frac{\sigma_{\text{tamaños}}}{\mu_{\text{tamaños}}}$ | 0-1 | Uniformidad en distribución |
| **Discriminación** | Promedio(Cramér's V top 20) | 0-1 | Poder separador de variables |
| **Coherencia** | Promedio(frecuencia modal) | 0-1 | Homogeneidad interna |
| **Puntaje Total** | $0.33\times\text{Balance} + 0.33\times\text{Discriminación} + 0.33\times\text{Coherencia}$ | 0-1 | Evaluación combinada |



## Evaluación de la Calidad de las Clusterizaciones

###  Distribución de Clusters
Todas las clusterizaciones presentan distribuciones asimétricas (clusters desbalanceados). En particular:

- **lca_k2**: 2 clusters (90.9% vs 9.1%)
- **lca_k3**: 3 clusters (62.1%, 30.2%, 7.7%)
- **lca_k4**: 4 clusters (58.9%, 29.4%, 9.4%, 2.2%)
- **lca_k5**: 5 clusters (59.3%, 24.8%, 7.1%, 6.5%, 2.3%)
- **lca_k6**: 6 clusters (56.9%, 17.1%, 13.9%, 6.4%, 4.1%, 1.7%)

**Problema estadístico**: Presencia de micro-clusters (menos del 5% de observaciones) que pueden no ser estadísticamente representativos.

###  Poder Discriminante de las Variables
- **Variables consistentemente discriminantes** (aparecen en el top 25 de todas las clusterizaciones): 
  pcd012, pcd013, pcd014, pcd015, pcd019, pcd020, pcd021, pcd027, pcd028, pcd029, pcd039, pal001, pni001, pmj001.
- **Variables con discriminación perfecta** (valor = 1.000) en al menos una clusterización: 
  pcd029, pod007, pmd012, pmd001, pmd008, pmd009.

###  Evolución Jerárquica (Estabilidad de Clusters)
- **Transiciones principales**:
  - lca_k2 a lca_k3: Cluster mayoritario (90.9%) se divide en dos (68.3% y 29.3%).
  - lca_k3 a lca_k4: Alta estabilidad en clusters mayoritarios (90.6% y 94.8%).
  - lca_k4 a lca_k5: Cluster mayoritario (96.5%) se mantiene en un solo cluster.
  - lca_k5 a lca_k6: Cluster mayoritario (95.9%) se mantiene.

### Patrones Combinados (Interacciones entre Variables)
- **Hallazgo crítico**: En la clusterización lca_k2, se encontraron 167 patrones de interacción significativos (diferencia > 0.15 en Cramér's V). De ellos, 151 (90.4%) son patrones con asociación más débil en los clusters comparado con el global.
- **Ejemplos destacados**:
  - pcd012 × pcd002: Asociación global = 0.711, en cluster = 0.011 (diferencia = -0.700).
  - pcd013 × pcd022: Asociación global = 0.710, en cluster = 0.015 (diferencia = -0.695).
  - pcd039 × pcd022: Asociación global = 0.714, en cluster = 0.040 (diferencia = -0.674).

**Interpretación**: Los clusters parecen estar definidos por combinaciones de categorías que rompen las asociaciones fuertes observadas a nivel global.

## Evaluación Comparativa Final

###  Puntuación por Clusterización (métrica compuesta)
1. lca_k3: 0.437
2. lca_k2: 0.385
3. lca_k4: 0.377
4. lca_k5: 0.324
5. lca_k6: 0.307

###  Componentes de la Puntuación
- **Balance**: Mejor en lca_k3 (0.330), peor en lca_k6 (-0.125).
- **Discriminación**: Aumenta con el número de clusters (de 0.309 a 0.372).
- **Coherencia**: Se mantiene estable (~0.67-0.68) en todas las clusterizaciones.

##  Conclusiones Clave

1. **Clusterización óptima**: **lca_k3** (3 clusters) es la que muestra el mejor equilibrio entre balance, discriminación y coherencia, además de tener un número manejable de clusters.

2. **Problema de desbalanceo**: Todas las clusterizaciones presentan clusters desbalanceados, con algunos micro-clusters (menos de 5% de observaciones) que requieren validación estadística adicional.

3. **Variables clave**: Existe un conjunto de variables (aproximadamente 15) que son consistentemente discriminantes en todas las clusterizaciones. Algunas muestran discriminación perfecta, lo que indica que sus categorías están perfectamente separadas entre clusters.

4. **Interacciones entre variables**: Los clusters se caracterizan por tener asociaciones más débiles entre variables que las observadas a nivel global. Esto sugiere que los clusters representan segmentos donde las relaciones poblacionales no se mantienen.

5. **Jerarquía estable**: Las transiciones entre clusterizaciones sucesivas muestran una alta estabilidad para los clusters mayoritarios, lo que valida la robustez de la estructura de clustering.



**Conclusión final**: El análisis revela una estructura de clustering válida pero con clusters desbalanceados. La clusterización lca_k3 (3 clusters) es la más recomendable por su equilibrio estadístico. Los resultados sugieren la existencia de segmentos bien diferenciados que operan bajo dinámicas distintas a las de la población general.

Ahora es necesario identificar y verbalizar los patrones encontrados en base a las preguntas del DISC-IV

In [14]:
resultados = analisis_estadistico_clusters(
    final_df=final_df,          
    id_column='id',              
    cluster_prefix='lca_k',     
    umbral_diferencia=0.10,      
    top_n_variables=25,         
    umbral_asociacion=0.15      
)

ANÁLISIS ESTADÍSTICO DE CLUSTERS

COLUMNAS DE CLUSTER ENCONTRADAS: ['lca_k2', 'lca_k3', 'lca_k4', 'lca_k5', 'lca_k6']
Total de observaciones: 1558

TOTAL variables para análisis: 136
Distribución por tipo de dato:
  category: 2 variables
  object: 134 variables

INICIANDO ANÁLISIS COMPLETO...

1. ESTADÍSTICAS DE DISTRIBUCIÓN DE CLUSTERS POR CLUSTERIZACIÓN

RESUMEN DE TODOS LOS CLUSTERS:

lca_k2:
  Número de clusters: 2
  Total observaciones: 1558
  Tamaño promedio de cluster: 779
  Rango de tamaños: 141 - 1417
  Completitud promedio: 99.2%

lca_k3:
  Número de clusters: 3
  Total observaciones: 1558
  Tamaño promedio de cluster: 519
  Rango de tamaños: 120 - 968
  Completitud promedio: 99.4%

lca_k4:
  Número de clusters: 4
  Total observaciones: 1558
  Tamaño promedio de cluster: 390
  Rango de tamaños: 35 - 918
  Completitud promedio: 98.6%

lca_k5:
  Número de clusters: 5
  Total observaciones: 1558
  Tamaño promedio de cluster: 312
  Rango de tamaños: 36 - 924
  Completitud promedi